### PR TITLE
fix issue 21149 - exclude templatized alias and aliasseq from -vtempl…

### PR DESCRIPTION
**Pull request by ghost**
…ates reports

Because their semantics are analyzed in a special way, that doesn't follow the standard path.
Also counting the instances in the special functions `aliasInstanceSemantic`
and `aliasSeqInstanceSemantic` would give wrong results (all uniques).

### CODE REVIEW

**Review by ibuclaw — APPROVED** [review]
Seems reasonable to me.

**Review by wilzbach — APPROVED** [review]
LGTM too

**Comment by UplinkCoder** [comment]
you could count them as something special?

**Comment by NilsLankila** [comment]
Yes.

**Comment by NilsLankila** [comment]
Maybe the names can be in the list but with a mention "(ignored because optimized)" or I don't know what. Or maybe a trailing paragraph after the table.

"template `thing`, `stuff`, `smurf` are ignored because of a special processing that optimize their instantiation"

**Comment by NilsLankila** [comment]
@UplinkCoder, I've added a second commit that prints the trailing paragraph that I mentioned.

File: test/compilable/b21149.d [hunk 1]
```diff
@@ -0,0 +1,27 @@
+/*
+REQUIRED_ARGS: -vtemplates
+TEST_OUTPUT:
+---
+  Number   Unique   Name
+
+template(s) `W(T...)`, `D(T)`, `P(T...)`, `F(T)`, `Y(T...)`, `L(T)`, `S(T...)`, `A(T)`,
+`Q(T...)`, `I(T)`, `Z(T...)`, `B(T)`, `E(T)`, `J(T)`, `M(T)`, `C(T)`, `T(T...)`, `X(T...)`,
+`V(T...)`, `U(T...)`, `G(T)`, `N(T...)`, `K(T)`, `R(T...)`, `O(T...)`, `H(T)`, are
+excluded because their instantiation follows an optimized path
```
Comment (NilsLankila) [inline test/compilable/b21149.d:10]: I'm afraid that the use of an asoc array might make this test failing randomly.

**Comment by wilzbach** [comment]
> Maybe the names can be in the list but with a mention "(ignored because optimized)" or I don't know what. Or maybe a trailing paragraph after the table.
> @UplinkCoder, I've added a second commit that prints the trailing paragraph that I mentioned.

I'm not sure how this will help the user. Isn't the primary purpose of `-vtemplates` to identify commonly used templated, i.e. a tool to reduce compilation cost. As self-templates are special-cased anyhow, I don't see a use cases for this, but it brings a few problems with it :/

If you want to keep that feature, I think it should at least be run against a big library like Phobos to see whether the output is actually useful.

**Comment by NilsLankila** [comment]
The point is that there might be complains about the fact that some templates are missing (bugzilla, new group), and each time somone will have to explain that something special happens in the compiler. This paragraph prevents this.

**Comment by wilzbach** [comment]
> The point is that there might be complains about the fact that some templates are missing (bugzilla, news group, IRC...),

You can never satisfy everyone. Anyhow, if you're worried, why not simply add a final line stating that identity templates are omitted from the list?

**Comment by nordlow** [comment]
> …ates reports
> 
> Because their semantics are analyzed in a special way, that doesn't follow the standard path.
> Also counting the instances in the special functions `aliasInstanceSemantic`
> and `aliasSeqInstanceSemantic` would give wrong results (all uniques).

Removing support for templated aliases removes an important part of the template diagnostics outputted by `-vtemplates`. Could we pleease leave this as is until we have a replacement solution for the unique count problem. The total number of templated aliases instances seem to be correct. And that is what I'm outputting in at https://github.com/dlang/dmd/pull/11463.

**Comment by andralex** [comment]
I don't know, @WalterBright should rule on this.

**Pull request closed without merging**